lib/broadway/producer.ex

```diff
@@ -236,4 +289,110 @@ defmodule Broadway.Producer do
             ":transformer option to transform produced events into message structs"
   end
+
+  defp maybe_rate_limit_and_buffer_messages(state, messages) do
+    if state.rate_limiting && messages != [] do
+      state = update_in(state.rate_limiting.message_buffer, &enqueue_batch(&1, messages))
+      rate_limit_and_buffer_messages(state)
+    else
+      {state, messages}
+    end
+  end
+
+  defp rate_limit_and_buffer_messages(%{rate_limiting: %{state: :closed}} = state) do
+    {state, []}
+  end
+
+  defp rate_limit_and_buffer_messages(%{rate_limiting: %{message_buffer: buffer}} = state) do
+    case RateLimiter.get_currently_allowed(state.rate_limiting.table_name) do
+      # No point in trying to emit messages if no messages are allowed. In that case,
+      # we close the rate limiting and don't emit anything.
+      allowed when allowed <= 0 ->
+        {put_in(state.rate_limiting.state, :closed), []}
+
+      allowed ->
+        {allowed_left, probably_emittable, buffer} = dequeue_many(buffer, allowed, [])
+
+        {rate_limiting_state, messages_to_emit, messages_to_buffer} =
+          rate_limit_messages(
+            state.rate_limiting.table_name,
+            probably_emittable,
+            _probably_emittable_count = allowed - allowed_left
+          )
+
+        new_buffer = enqueue_batch_r(buffer, messages_to_buffer)
+        state = put_in(state.rate_limiting.message_buffer, new_buffer)
+        state = put_in(state.rate_limiting.state, rate_limiting_state)
+
+        {state, messages_to_emit}
+    end
+  end
+
+  defp reverse_split_demand(rest, 0, acc) do
+    {0, acc, rest}
+  end
+
+  defp reverse_split_demand([], demand, acc) do
+    {demand, acc, []}
+  end
+
+  defp reverse_split_demand([head | tail], demand, acc) do
+    reverse_split_demand(tail, demand - 1, [head | acc])
+  end
+
+  defp dequeue_many(queue, demand, acc) do
+    case :queue.out(queue) do
+      {{:value, list}, queue} ->
+        case reverse_split_demand(list, demand, acc) do
+          {0, acc, []} ->
+            {0, Enum.reverse(acc), queue}
+
+          {0, acc, rest} ->
+            {0, Enum.reverse(acc), :queue.in_r(rest, queue)}
+
+          {demand, acc, []} ->
+            dequeue_many(queue, demand, acc)
+        end
+
+      {:empty, queue} ->
+        {demand, Enum.reverse(acc), queue}
+    end
+  end
+
+  defp enqueue_batch(queue, _list = []), do: queue
+  defp enqueue_batch(queue, list), do: :queue.in(list, queue)
+
+  defp enqueue_batch_r(queue, _list = []), do: queue
+  defp enqueue_batch_r(queue, list), do: :queue.in_r(list, queue)
+
+  defp rate_limit_messages(_state, [], _count) do
+    {[], []}
+  end
+
+  defp rate_limit_messages(table_name, messages, message_count) do
+    case RateLimiter.rate_limit(table_name, message_count) do
+      # If no more messages are allowed, we're rate limited but we're able
+      # to emit all messages that we have.
+      0 ->
+        {:closed, messages, _to_buffer = []}
```

Comment:
We can probably remove this clause and merge it with the last clause. `Enum.split(messages, 0)` is basically a no-op for lists.

Reply:
`Enum.split(messages, 0)` returns `{[], ...}` but when the rate limit left is 0, we want to return `{messages, []}`. When it's `overflow when overflow < 0`, we use the trick that `Enum.split` works with negative indexes so that `Enum.split(messages, -5)` will return `{messages_to_send, messages_that_overflow}` correctly. Makes sense?

Reply:
Ah yes, thanks!